Scheduled toot

kyori19 opened the pull request:
Related to #981 .

I added button to configure scheduled status and add `scheduled_at` param to `createStatus` method.

But there seems to be a problem with scheduled toots contains images.(tested on Mastodon v2.7.0 master)
This is not a bug of this PR.

My toot posted with this feature. -> https://odakyu.app/@ars42525/101455089752932693
![](https://media.odakyu.app/media_attachments/files/000/341/573/small/7aef570b6f87d81b.png)

![](https://media.odakyu.app/media_attachments/files/000/341/574/small/6606b5ae94e1a41b.png)

![](https://media.odakyu.app/media_attachments/files/000/341/575/small/10e79dedd90b8899.png)

charlag commented:
Woah, thanks, that would be super cool to have!
My biggest concern so far is UI-wise. I'll try to go through the code during the week.

charlag commented on app/src/main/java/com/keylesspalace/tusky/fragment/DatePickerFragment.java:
```diff
@@ -0,0 +1,41 @@
+
+public class DatePickerFragment extends DialogFragment {
+
+    public static final String PICKER_TAG = "date_picker";
```
Do we need this tag?

charlag commented on app/src/main/java/com/keylesspalace/tusky/network/MastodonApi.java:
```diff
@@ -33,6 +31,7 @@
 
 import java.util.List;
 
+import androidx.annotation.Nullable;
```
Please remove unused imports in few places

kyori19 replied:
I think there's no unused imports in this PR.
Am I missing something?

charlag replied:
Maybe I'm reading incorrectly on Github, sorry

kyori19 commented:
* Moved `PICKER_TAG` to `picker.show(fragmentManager, tag)`
* Added `VersionUtils` to check Mastodon version and hide schedule button if it is lower than 2.7.0

I'm not good at writing Kotlin. Can I use Android Studio's "migrate to Kotlin" feature?

charlag commented:
I usually convert existing things automatically and then fix rough edges. We can try to do this if you want but if you would rather not then just leave it like that.

connyduck commented:
- The date/time picker dialog is ugly as hell, because its not from the material design library. Coincidentally, the Android team just started to work on one. I guess we have to wait until they release it. https://github.com/material-components/material-components-android/blob/704aff2f14d3b985ab35fe83e22d239c08caf943/lib/java/com/google/android/material/picker/MaterialDatePickerDialog.java

- Nice idea with the bottom sheet. I would change the behavior so that it hides after a time was selected, but the clock icon turns blue to indicate the toot will be sent scheduled.

- There should be an icon next to the set time/date to indicate its editable.

- There has to be a way to remove the time and send the status immediately.

kyori19 commented:
I found problem with timeline auto-reloading after toot so I fixed it and rebased branch.

charlag commented:
@kyori19 sorry that it takes too long, we'll try to get back to it whenever possible!
You need a new review at this stage, right?

kyori19 commented:
@charlag 
Thanks, I'm sure I need some reviews to improve this feature. And I have to make change the UI, so I will work on it in near.

kyori19 commented:
I improved bottom sheet design.

Added edit button, added reset button and schedule button turns blue when a time is set.

![](https://media.odakyu.app/media_attachments/files/000/522/068/original/a50a24b074258e36.jpg)

charlag commented:
I would revese right and left part of the sheet and maybe made reset button less prominent.
Edit icon doesn't seem aligned with text. But nice progress!
(I'm not maintainer and I just leave my feedback, feel free to ignore it)

kyori19 commented:
@charlag Thanks!
I will fix about the edit icons alignment sooner.
I think it is better that the edit icons are placed on the right side of text.
If the reset button placed on the right of the sheet, strange space will appear on left of the edit icon.
But I'm also one of contributor, so I want to hear @connyduck's opinion.

kyori19 commented:
Fixed edit icon's size.
![image](https://user-images.githubusercontent.com/27999567/54485380-d5d2f500-48ba-11e9-9567-c81cd0717fee.png)

kyori19 commented:
Added a list of scheduled toots to drawer.
There's no API to edit scheduled status so once delete it and show `ComposeActivity` when edit statuses.

![image](https://user-images.githubusercontent.com/27999567/54492282-78659500-4908-11e9-962d-6dc938cc3c1b.png) ![image](https://user-images.githubusercontent.com/27999567/54492284-7f8ca300-4908-11e9-8b72-f9ea760dbef1.png)

kyori19 commented on app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java:
```diff
@@ -1152,7 +1196,14 @@ private void addMediaToQueue(@Nullable String id, QueuedMedia.Type type, Bitmap
+        if (preview != null) {
+            view.setImageBitmap(preview);
+        } else {
+            Picasso.with(this)
```
Use picasso to show media preview when it has been already uploaded.

charlag commented:
@kyori19 some things have been changed in master. Would you be willing to update it?

kyori19 commented:
@charlag 
Sorry for waiting you long.
Rebased so that you can merge.
Furthermore, I fixed wrong version detection. 

And I am still waiting for completion of MaterialDateTimePicker so please wait...

charlag commented:
Ah, sorry, I thought you've been waiting for us to merge it and it's getting older.
Good that you rebased it anyway, it's easier to do it regularly. Thanks.

connyduck commented:
Looks like the newest Materialcomponents library has an updated public MaterialDatePickerDialog https://github.com/material-components/material-components-android/releases/tag/1.1.0-alpha10

Can you resolve the merge conflicts and try it out please? @kyori19 
Let me know if I can help with something

kyori19 commented:
@connyduck 
Yeah, I know. I already resolved all conflicts and working on implementing with new datepicker.
It seems that I need few days to complete this, so please wait...

kyori19 commented:
@connyduck I changed the design of ScheduleView.
I think the strange parts has been improved.
Please check them.

kyori19 commented:
Resolved all conflicts.
